editor/project_manager.cpp

```diff
@@ -69,270 +70,303 @@ constexpr int GODOT4_CONFIG_VERSION = 5;
+			if (!pkg) {
+				_set_message(TTR("Invalid \".zip\" project file; it is not in ZIP format."), MESSAGE_ERROR);
+				unzClose(pkg);
+				return "";
+			}
 
-					ret = unzGoToNextFile(pkg);
+			int ret = unzGoToFirstFile(pkg);
+			while (ret == UNZ_OK) {
+				unz_file_info info;
+				char fname[16384];
+				ret = unzGetCurrentFileInfo(pkg, &info, fname, 16384, nullptr, 0, nullptr, 0);
+				if (ret != UNZ_OK) {
+					break;
 				}
 
-				if (ret == UNZ_END_OF_LIST_OF_FILE) {
-					_set_message(TTR("Invalid \".zip\" project file; it doesn't contain a \"project.godot\" file."), MESSAGE_ERROR);
-					get_ok_button()->set_disabled(true);
-					unzClose(pkg);
-					return "";
+				if (String::utf8(fname).ends_with("project.godot")) {
+					break;
 				}
 
+				ret = unzGoToNextFile(pkg);
+			}
+
+			if (ret == UNZ_END_OF_LIST_OF_FILE) {
+				_set_message(TTR("Invalid \".zip\" project file; it doesn't contain a \"project.godot\" file."), MESSAGE_ERROR);
 				unzClose(pkg);
+				return "";
+			}
 
-				// check if the specified install folder is empty, even though this is not an error, it is good to check here
-				d->list_dir_begin();
-				is_folder_empty = true;
-				String n = d->get_next();
-				while (!n.is_empty()) {
-					if (!n.begins_with(".")) {
-						// Allow `.`, `..` (reserved current/parent folder names)
-						// and hidden files/folders to be present.
-						// For instance, this lets users initialize a Git repository
-						// and still be able to create a project in the directory afterwards.
-						is_folder_empty = false;
-						break;
-					}
-					n = d->get_next();
-				}
-				d->list_dir_end();
+			unzClose(pkg);
+		} else {
+			install_path_container->hide();
+			create_dir->hide();
+			_set_message(TTR("The path specified is not a valid project."), MESSAGE_ERROR);
+			return "";
+		}
 
-				if (!is_folder_empty) {
-					_set_message(TTR("Please choose an empty folder."), MESSAGE_WARNING, INSTALL_PATH);
-					get_ok_button()->set_disabled(true);
-					return "";
-				}
+		target_path = install_path->get_text().simplify_path();
+		target_path_input_type = INSTALL_PATH;
+	}
 
-			} else {
-				_set_message(TTR("Please choose a \"project.godot\", a directory with it, or a \".zip\" file."), MESSAGE_ERROR);
-				install_path_container->hide();
-				get_ok_button()->set_disabled(true);
+	String working_dir = d->get_current_dir();
+	String executable_dir = OS::get_singleton()->get_executable_path().get_base_dir();
+	if (target_path == working_dir || target_path == executable_dir) {
+		_set_message(TTR("Creating a project in the engine's working directory or executable directory is not allowed, as it would prevent the project manager from starting."), MESSAGE_ERROR, target_path_input_type);
+		return "";
+	}
+
+	// TODO: The following 5 lines could be simplified if OS.get_user_home_dir() or SYSTEM_DIR_HOME is implemented.
+	// Proposal: https://github.com/godotengine/godot-proposals/issues/4851
```

Comment:
TODO comment is ok, proposal link is out of place IMO.

Reply:
Changed to:
```cpp
// TODO: The following 5 lines could be simplified if OS.get_user_home_dir() or SYSTEM_DIR_HOME is implemented. See: https://github.com/godotengine/godot-proposals/issues/4851.
```